### PR TITLE
Add Push doc

### PR DESCRIPTION
**Pull request by danielpassos**

### CODE REVIEW

**Comment by wtrocki** [comment]
@danielpassos Copy paste worked.

File: docs/modules/getting-started/pages/push.adoc [hunk 1]
```diff
@@ -0,0 +1,112 @@
...
+
+For an example of push configuration see link:../../example/src/main/assets/mobile-services.json[example apps mobile-services.json].
+
+== Registering with Unified Push Server
+
+This example demonstrates how to register a device to provide the values required to register the application with FCM and Unified Push and perform the registration.
+
+[source,java]
+----
+PushService pushService = MobileCore.getInstance().getService(PushService.class);
```
[the remark this answers is not in the document]
Comment (danielpassos) [reply docs/modules/getting-started/pages/push.adoc]: There is not the best place to add that since it's a singleton... you can do it when wherever you want how many times you want ;)

[the remark this answers is not in the document]
Comment (wtrocki) [reply docs/modules/getting-started/pages/push.adoc]: I know but noob doesn't know that and first thing he is thiking. Where I should put this stuff. Application or Activity onCreate is the best place probably.

File: docs/modules/getting-started/pages/push.adoc [hunk 2]
```diff
@@ -0,0 +1,112 @@
...
+Push messaging is a mechanism for sending data to mobile clients from a remote server. For mobile applications, this reduces battery usage, removes polling code and provides for a better user experience as data changes are communicated in real time.
+
+Messages that FCM receives from an application server are sent to the Android device over a connection managed by the operating system. These messages are then extracted and passed to an application using services definied provided by the AeroGear library.
+
+== Adding dependency
+
+Add the push dependency to your application module
+
+----
+implementation "org.aerogear:aerogear-push:[version]"
```
[the remark this answers is not in the document]
Comment (wtrocki) [reply docs/modules/getting-started/pages/push.adoc]: I was following the guide. Seems to be not updated.

https://github.com/aerogear/aerogear-android-sdk/blob/master/docs/modules/getting-started/pages/getting-started.adoc#bill-of-materials

Do not get me wrong. I'm really happy to see BOM go away - it's just still there.

[the remark this answers is not in the document]
Comment (danielpassos) [reply docs/modules/getting-started/pages/push.adoc]: @wtrocki I see. Most of the docs are outdated. My next step will be updating all docs on Android SDK, to reflect the new changes not documented. I was only waiting to get @finp PR (https://github.com/aerogear/aerogear-android-sdk/pull/204) merged

[the remark this answers is not in the document]
Comment (secondsun) [reply docs/modules/getting-started/pages/push.adoc]: I don't think we've released the SDK since the bom was taken out so those docs aren't worthless yet.

But yeah we should flag an issue to update the docs

[the remark this answers is not in the document]
Comment (danielpassos) [reply docs/modules/getting-started/pages/push.adoc]: @secondsun I will review/update all Android SDK docs after got it merged

File: docs/modules/getting-started/pages/push.adoc (hunk 1 above)
[the remark this answers is not in the document]
Comment (secondsun) [reply docs/modules/getting-started/pages/push.adoc]: There are several places registration can happen.  If you want to always be registered then you add it in the Application class.  If you want users to have control over it then you need to add it into a Activity.  There are reasons for both and those reasons are different and nuanced.

I think that the best place to introduce people to it is in an event handle on the activity (ie handleRegisterButtonPress or something) because we want to get peoeple used to the idea that push messages are opt in.

TL;DR; wrap it in an activity in the code sample and have register and unregister be in methods that are event handlers, thoughts?

File: docs/modules/getting-started/pages/push.adoc [hunk 3]
```diff
@@ -0,0 +1,112 @@
...
+PushService pushService = MobileCore.getInstance().getService(PushService.class);
+pushService.unregisterDevice(new Callback() {
+    @Override
+    public void onSuccess() {
+        // yay
+    }
+
+    @Override
+    public void onError(Throwable error) {
+        // Ops!
```
Comment (secondsun) [inline docs/modules/getting-started/pages/push.adoc]: /s/Ops/Oops.

**Comment by wtrocki** [comment]
@danielpassos Checked that again! Looks really good!

File: docs/modules/getting-started/pages/push.adoc [hunk 4]
```diff
@@ -0,0 +1,112 @@
...
+    }
+});
+----
+
+== Handling messages
+
+`MessageHandler` is responsable for handling the messages received from the FCM. It will give you total flexibility to decide what to do with the message recieved.
```
Comment (secondsun) [inline docs/modules/getting-started/pages/push.adoc]: "from FCM"

File: docs/modules/getting-started/pages/push.adoc [hunk 5]
```diff
@@ -0,0 +1,112 @@
...
+
+## Receiving Background and Foreground messages
```
Comment (secondsun) [inline docs/modules/getting-started/pages/push.adoc]: capitalize messages

File: docs/modules/getting-started/pages/push.adoc [hunk 6]
```diff
@@ -0,0 +1,112 @@
...
+== Default handler
+
+Unfortunately when you close or kill the app the Android system can destroy the `PushService` instance and we will lose the handler you have added there. To solve this we provide a way to register a default handler, it will be used when app is not running (actually running in background) and there are no others handlers registered.
+
+[source,xml]
+----
+<meta-data
+    android:name="DEFAULT_MESSAGE_HANDLER_KEY"
+    android:value="package.ClassName" />
+----
```
Comment (secondsun) [inline docs/modules/getting-started/pages/push.adoc]: There are two details missing.  1) This stanza has to be in the application block and 2) this has to be in AndroidManifest.xml

File: docs/modules/getting-started/pages/push.adoc (hunk 1 above)
[the remark this answers is not in the document]
Comment (danielpassos) [reply docs/modules/getting-started/pages/push.adoc]: I got you both! My point is... IMO we should not mention in the doc _where_  developer should do that and let he/she free to do it wherever he/she wants

@secondsun The activity example you mention is exactly what we have in the example app. The developer can always consult it ;)

**Comment by danielpassos** [comment]
I just realized I forgot to metion `google-service.json`. Going to add it

**Comment by danielpassos** [comment]
@wtrocki @secondsun I have added a new topic about Firebase setup and also create a new project from scratch using this step by step and it works like a charm.

Can you please double check it?

**Comment by wtrocki** [comment]
@secondsun Could we merge that. It seems to be clear now.

**Comment by wtrocki** [comment]
Done final run. 

Typos:
> `MessageHandler` is responsable for handling the messages received from FCM. It will give you total flexibility to decide what to do with the message recieved.

> ## Receiving Background and Foreground Messages

Non asci docs style.

File: docs/modules/getting-started/pages/push.adoc [hunk 7]
```diff
@@ -0,0 +1,144 @@
...
+        // Oops!
+    }
+});
+----
+
+== Handling messages
+
+`MessageHandler` is responsable for handling the messages received from FCM. It will give you total flexibility to decide what to do with the message recieved.
+
+## Receiving Background and Foreground Messages
```
Comment (wtrocki) [inline docs/modules/getting-started/pages/push.adoc]: Non ascii style

File: docs/modules/getting-started/pages/push.adoc [hunk 8]
```diff
@@ -0,0 +1,144 @@
...
+    @Override
+    public void onError(Throwable error) {
+        // Oops!
+    }
+});
+----
+
+== Handling messages
+
+`MessageHandler` is responsable for handling the messages received from FCM. It will give you total flexibility to decide what to do with the message recieved.
```
Comment (wtrocki) [inline docs/modules/getting-started/pages/push.adoc]: responsable  and recieved

**Comment by danielpassos** [comment]
@wtrocki fixed!

File: docs/modules/getting-started/pages/push.adoc (hunk 1 above)
[the remark this answers is not in the document]
Comment (secondsun) [reply docs/modules/getting-started/pages/push.adoc]: @danielpassos Sounds good.
